src/js/shape.js

```diff
@@ -0,0 +1,72 @@
+import Block from './block';
+import { CONSTANTS } from './values';
+
+export default class Shape {
+  constructor({ x, y, container, rotation, unitSize }) {
+    this.x = x;
+    this.y = y;
+    this.container = container;
+    this.rotation = rotation;
+    this.unitSize = unitSize;
+    this.blocks = [];
+  }
+
+  draw(container) {
+    let options = this.constructor.blockOptions[this.rotation];
+    let blocks = options.map(point => new Block({
+      x: point.x + this.x,
+      y: point.y + this.y,
+      unitSize: this.unitSize,
+      color: this.constructor.color
+    }));
+    blocks.forEach(block => {
+      block.draw(container);
+    });
+    this.blocks = blocks;
+  }
+
+  clone() {
+    return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
+  }
+
+  canMoveLeft(oldBlocks) {
+    return this.blocks.every(block => block.x !== 0)
+      && oldBlocks.every(block => !this.blocks.some(b => (b.x - 1) === block.x && b.y === block.y));
+  }
+
+  canMoveRight(oldBlocks) {
+    return this.blocks.every(block => block.x !== (CONSTANTS.tetrisWidth - 1))
+      && oldBlocks.every(block => !this.blocks.some(b => (b.x + 1) === block.x && b.y === block.y));
+  }
+
+  canMoveDown(oldBlocks) {
+    return this.blocks.every(block => block.y !== ((CONSTANTS.tetrisWidth * 2) - 1))
+      && oldBlocks.every(block => !this.blocks.some(b => b.x === block.x && (b.y + 1) === block.y));
+  }
```

Comment:
We can clean here a bit by doing something like
```js
moveDown() {
  return this.simulateMove(() => 
    this.y += 1
  );
}

moveLeft() {
  return this.simulateMove(() => 
    this.x -= 1;
  );
}


moveRight() {
  return this.simulateMove(() => 
    this.x += 1;
  );
}
```

simulateMove here can draw the next move, check if it hits, if it does it redraws the previous one.